### PR TITLE
Added hooks `method_added` and `method_removed`

### PR DESCRIPTION
**Pull request by MakeNowJust**
Please see #1065. This is implementation of it.

If you have an argument of `method_added` and `method_removed` specification, please post a comment to #1065. But if you have an argument of this implementation, please post there.

### CODE REVIEW

**Comment by asterite** [comment]
@MakeNowJust Awesome!! I don't have time to take a look at this, but I'll do it during this week. It's amazing that you modified `type_inference.cr`, I don't think that code is very clear, but you found a way :-)

I think `undef` will be gone, so maybe `method_removed` won't be needed.

I read in crystal-jp that after this change, the compiler takes 7s more to compile. Did you try compiling the new compiler with `--release`? If it's slow, maybe it's because you always create a `Call` node, even if there are no hooks, so maybe creating it lazily will speed up things (but I don't want to guess much, profiling is bette)

**Comment by MakeNowJust** [comment]
@asterite  I wrote [benchmark script and run it](https://gist.github.com/MakeNowJust/1cc3acc531eda2ad63af).

According this result, added `method_added` version is faster rather than old version.  Why? Please check benchmark script.

**Comment by asterite** [comment]
@MakeNowJust Oh, to check the time it takes to compile the compiler, we usually don't use the `--release` flag. It's a bit confusing:

1. Compile the new compiler using the `--release` flag, so the compiler is optimized.
2. With the new compiler, compiler another compiler **not using** the `--release` flag.

The problem is, `--release` will turn on LLVM optimizations, and this is where most of the time is spent (about a minute or more). 

To make it clear:

```
Parse: 00:00:00.1067890               # to parse the whole program
Type inference: 00:00:05.3825190      # the whole type checking, we are ready for codegen after this
Codegen (crystal): 00:00:01.6943680   # generate LLVM IR code
Codegen (bc+obj): 00:00:08.1608280    # turn that LLVM IR code into .o files (!!)
Codegen (clang): 00:00:01.0178690     # run `clang ...` with all the .o files
```

In the above stats you can see that 8 seconds are gone turning LLVM IR code into .o files (the `!!` part). And that is taking more than anything else we are doing (parsing and type inference). The next time you compile your program we try to reuse those `.o` files if they didn't change, so that time is usually reduced to zero:

```
# second compile, no change
Parse: 00:00:00.1095250
Type inference: 00:00:04.1833090
Codegen (crystal): 00:00:01.7189030
Codegen (bc+obj): 00:00:00.8762690 # time is reduced A LOT
Codegen (clang): 00:00:01.0008120
```

You can also see the second part, "type inference" is a bit faster too. That's because the compiler uses the "ecr" run macro, and that program is already compiled so there's no need to recompile it (we actually compile it, but the .o files are already there so there's no need to regenerate them... so... it's recursive).

Now, I compile with `--release`:

```
Parse: 00:00:00.1037790
Type inference: 00:00:05.0788610
Codegen (crystal): 00:00:01.5548660
Codegen (bc+obj): 00:03:59.5582060   # <--- 4 minutes!!
Codegen (clang): 00:00:00.6108890
```

Now it takes 4 minutes for that long phase, because before generating the .o files we optimize the LLVM IR.

Because of this, if you make changes to the type inference phase and want to check difference in performance, you have to check the second phase only, because the 4 minutes phase sometimes takes 4.05s, sometimes 3.55s, etc. (easier to small relative variations here).

As a closing comment, I tried your branch and didn't see any impact in the performance of the type inference :-)

**Comment by MakeNowJust** [comment]
@asterite Thanks for detailed comment!

I am going to remove `method_removed`, and I will add it again if decided `undef` alternative. OK?

BTW, how do you show those seconds?

**Comment by asterite** [comment]
@MakeNowJust Oh, sorry! I removed the first line. You use the `--stats` or `-s` flag:

```
$ bin/crystal build samples/mandelbrot.cr -s
Parse: 00:00:00.0317140
Type inference: 00:00:00.0652610
Codegen (crystal): 00:00:00.0188620
Codegen (bc+obj): 00:00:00.0841730
Codegen (clang): 00:00:00.0562710
```

**Comment by asterite** [comment]
@MakeNowJust You can leave `method_removed` for now. If we later decide to remove `undef`, we also remove that hook.

**Comment by MakeNowJust** [comment]
@asterite OK. I leave `method_removed` for now.

**Comment by MakeNowJust** [comment]
@asterite When this PR is merged? Or you have a reason not to merge it?

**Comment by jhass** [comment]
You might want to rebase to keep it easily mergable, right now there are conflicts ;)

**Comment by MakeNowJust** [comment]
@jhass Ok. I rebased!

**Comment by asterite** [comment]
@MakeNowJust Thanks! We'll probably merge this. But what are the uses cases for this? If you can't have state in macros, what would you use `method_added` for?

**Comment by jhass** [comment]
Since `included` can't catch all method definitions, especially when reopening classes no matter how we change it, it can be useful to dynamically wrap all method definitions for some instrumentation for example. Or sometimes the state doesn't need to be available at compile time and you can do very well build runtime state in macros.

**Comment by MakeNowJust** [comment]
@jhass Thanks! I considered same problems and examples too.

**Comment by asterite** [comment]
@MakeNowJust Could you rebase this pull request? Many things have changed, for example "undef" was removed and I also changed some arguments order for "expand_macro". After that we'll merge this. Thanks!

**Comment by MakeNowJust** [comment]
@asterite I rebased this branch.  Ok?

File: src/compiler/crystal/semantic/ast.cr [hunk 1]
```diff
@@ -628,6 +628,14 @@ module Crystal
     include RuntimeInitializable
   end
 
+  class Def
+    include RuntimeInitializable
+  end
+
+  class Undef
+    include RuntimeInitializable
+  end
```
Comment (jhass) [inline src/compiler/crystal/semantic/ast.cr]: Leftover?

File: src/compiler/crystal/types.cr [hunk 2]
```diff
@@ -627,6 +627,10 @@ module Crystal
         return add_hook :included, a_def
       when "extended"
         return add_hook :extended, a_def
+      when "method_added"
+        return add_hook :method_added, a_def, args_size: 1
+      when "method_removed"
+        return add_hook :method_removed, a_def, args_size: 1
```
Comment (jhass) [inline src/compiler/crystal/types.cr]: With undef gone, this is no longer useful, right?

**Comment by MakeNowJust** [comment]
Thanks @jhass.  I removed all `Undef` from this branch.

**Comment by jhass** [comment]
LGTM, but I think it's fine as a single commit. Could you squash them together please?

```
git remote add upstream https://github.com/manastech/crystal.git
git fetch upstream
git checkout feature/method_added
git rebase -i upstream/master
# Choose fix or squash for all but the first commit
git push -f origin feature/method_added # assuming origin is your fork

**Comment by jhass** [comment]
Thank you!

**Pull request merged**